AsyncDisplayKit/ASDisplayNode.mm

```diff
@@ -1912,13 +1912,7 @@ - (ASLayout *)calculateLayoutThatFits:(ASSizeRange)constrainedSize
                                                size:layout.size
                                          sublayouts:@[layout]];
     }
-    return [layout flattenedLayoutUsingPredicateBlock:^BOOL(ASLayout *evaluatedLayout) {
-      if (self.usesImplicitHierarchyManagement) {
-        return ASObjectIsEqual(layout, evaluatedLayout) == NO && [evaluatedLayout.layoutableObject isKindOfClass:[ASDisplayNode class]];
-      } else {
-        return [_subnodes containsObject:evaluatedLayout.layoutableObject];
```

Comment:
Turns out you can't do a simple check of whether the node is a subnode to determine if it should be in the flattened layout tree. Why? Because when implicit hierarchy management is turned on, the layout pass is performed before the nodes are even inserted into the hierarchy. The solution here is to do a simple identity check instead and use the `flattened` ivar to determine if a particular subtree has already been visited. Visited subtrees belong to subnodes of the current node, which turns out to be a good way to determine that a node isn't a direct subnode.